composer.json

```diff
@@ -34,7 +34,8 @@
         "amphp/socket": "^1",
         "amphp/sync": "^1.3",
         "league/uri": "^6",
-        "psr/http-message": "^1"
+        "psr/http-message": "^1",
+        "psr/http-factory": "^1"
```

Comment:
now you can also remove this one as it's not a dependency at all!
(it's only a dependency of the external part that instantiate the factory)

Reply:
It's a dependency as much as `psr/http-message` is a dependency for `UriInterface`.

Reply:
Look at the code: `UriInterface` is part of the public signature of `Request::getUri()`.

`RequestFactoryInterface` on the contrary is only part of the constructor of `PsrAdapter`. This means you *can* load the class definition of `PsrAdapter` without any issue. (And you can't for `Request`).

The thing that instantiates `PsrAdapter` must have `psr/http-factory` as a dependency to use the class, but the class and the package themself do not dependent on the `psr/http-factory`.

Reply:
I agree that `psr/http-factory` is much lesser evil than Laminas - it's just interfaces. In case it's unacceptable - the only escape is to extract adapters to separate library.

Reply:
But it is *not* dependency of this package... This should be moved to require-dev as that's the place were it belongs.

Reply:
Sorry, but it _is_ dependency, because if we use type in our signature (no matter if as argument or result); as a prrof of my words - require-checker fails without it.

Reply:
Then that's a bug in require-checker... If you reason about what makes something a dependency, this one is not. Because it's used only in the constructor, it's for sure a dependency of the thing that is passed to the constructor, but it's not a dependency of this specific class.

Trimming what is in "require" is critical to escape from dependency hell, here is an opportunity that we should not miss.

Reply:
No, it's not a bug. Look: this `RequestFactoryIntreface` is not just an obscure type; we call it's methods and thus we're depending on it's _version_.

I absolutely agree that not every user of `amphp/http-client` needs these adapters at all; we can extract them into the separate library, something like `amphp/http-client-psr7-bridge`. In the same time, supporting a separate library is a work; and PSR packages are small and change rarely. So I delegate the decision to @kelunik - his code, his rules.

Reply:
amphp/http-client should not force a dependency on its dependent packages when it's not *required* to have it. This line pushes a dependency that will be dead code to ppl that don't use `PsrAdapter`. And ppl that will use `PsrAdapter` *must* depend on `psr/http-factory` on their own because the are the one that will instantiate that injected object.

If you really want to enforce version boundaries, this should be moved to "conflict" instead.

Reply:
> This line pushes a dependency that will be dead code to ppl that don't use `PsrAdapter`.

I absolutely agree. But there is no way to detect what parts of a package are _really_ used; thus the only way we have is just assuming that _any_ part of library can be used. Package have to stay consistent.

Is I've said above, the only way to remove this dependency is to move the adapter to another package. "Conflict" is not an option.

Reply:
Let's say we remove this line. What did we lose in the process? Did we break anything? What does this line provide in the first place?

Reply:
@remorhaz: "conflict" is very much an option, and probably the way to go.

As most users won't use the PSR converter, I'm leaning towards a separate package, as breaking changes seem more likely than for the rest of the code in this package.

Reply:
> Let's say we remove this line. What did we lose in the process? Did we break anything?

Of course. If someone uses factory v1 in his code and then upgrades to factory v2 that breaks BC (by renaming factory method, for example), he gets his service broken without any chance of being preserved from it by Composer. That's how the dependency works; just ignoring the dependency doesn't remove it.

"Conflicting" with factory v2+ is pretty the same as requiring v1 - it will not allow to install factory v2 with our library, even if user doesn't use PSR adapters. It's just using wrong mechanism instead of right one, I think; it solves no problems. I hope @kelunik will decide to create a separate package - that's the best way to keep original client's dependencies untouched.

Reply:
Software design, package principles here, should prepare us for events that are possible. Here the v1/v2 example is unrealistic IMHO when talking about `psr/http-factory`. It's not a good enough reason to force it as a dependency to everyone "just in case". Even adding a "conflict" looks like useless precaution to me.

Anyway, we don't have to agree on this as a separate package might provide a path forward for both our pov.

Thanks for the talk.

Reply:
@nicolas-grekas In the case of `psr/http-factory` the BC break is indeed unlikely, as PSRs don't currently evolve and probably never will, because releasing a v2 on the same package would be a huge pain.

"conflict" seems to be the right choice currently for we're compatible with this API, but don't want an installation dependency.

Indeed, a separate package solves the "problem" entirely. :)